tests/processing/strategies/test_streaming.py

```diff
@@ -373,38 +377,96 @@ def test_parallel_transform_worker_apply() -> None:
     output_block = smm.SharedMemory(4096)
     assert output_block.size == 4096
 
-    index, output_batch = parallel_transform_worker_apply(
+    result = parallel_transform_worker_apply(
         transform_payload_expand,
         input_batch,
         output_block,
     )
 
     # The first batch should be able to fit 2 messages.
-    assert index == 2
-    assert len(output_batch) == 2
+    assert result.next_index_to_process == 2
+    assert len(result.valid_messages_transformed) == 2
 
-    index, output_batch = parallel_transform_worker_apply(
+    result = parallel_transform_worker_apply(
         transform_payload_expand,
         input_batch,
         output_block,
-        index,
+        result.next_index_to_process,
     )
 
     # The second batch should be able to fit one message.
-    assert index == 3
-    assert len(output_batch) == 1
+    assert result.next_index_to_process == 3
+    assert len(result.valid_messages_transformed) == 1
 
     # The last message is too large to fit in the batch.
     with pytest.raises(ValueTooLarge):
         parallel_transform_worker_apply(
             transform_payload_expand,
             input_batch,
             output_block,
-            index,
+            result.next_index_to_process,
         )
     smm.shutdown()
 
 
+NO_KEY = "No Key"
+
+
+def fail_bad_messages(message: Message[KafkaPayload]) -> KafkaPayload:
+    if message.payload.key is None:
+        raise InvalidMessages(
+            [
+                InvalidKafkaMessage(
+                    payload=str(message.payload),
+                    timestamp=message.timestamp,
+                    topic=message.partition.topic.name,
+                    consumer_group="",
+                    partition=message.partition.index,
+                    offset=message.offset,
+                    headers=message.payload.headers,
+                    reason=NO_KEY,
+                )
+            ]
+        )
+    return message.payload
+
+
+def test_parallel_transform_worker_bad_messages() -> None:
+    smm = SharedMemoryManager()
+    smm.start()
+    input_block = smm.SharedMemory(128)
+    output_block = smm.SharedMemory(128)
+    input_batch = MessageBatch[Any](input_block)
+
+    # every other message has a key
+    messages = [
+        Message(
+            Partition(Topic("test"), 0),
+            i,
+            KafkaPayload(None if i % 2 == 0 else b"key", b"\x00", []),
+            datetime.now(),
+        )
+        for i in range(9)
+    ]
+
+    for message in messages:
+        input_batch.append(message)
+
+    assert len(input_batch) == 9
+
+    result = parallel_transform_worker_apply(
+        fail_bad_messages,
+        input_batch,
+        output_block,
+    )
+    # all 9 messages processed
+    assert result.next_index_to_process == 9
+    # 5 were bad, 4 were good
+    assert len(result.invalid_messages) == 5
+    assert len(result.valid_messages_transformed) == 4
+    smm.shutdown()
```

Comment:
Please test the case where `start_index` is not 0. That is the interesting case to test and to cover.